# Notebook E-tivity 1 CE4021

**Student name:** Guillermo Alcantara Gonzalez

**Student ID:** 23123982

<hr style="border:2px solid gray"> </hr>

# 1: Derivatives

## Imports

In [1]:
from CE4021.Calculus import (
    derive_poly,
    evaluate_polynomial,
    integral,
    riemann_integral_approximation
)

## derive_poly

### Use example:
Let's calculate the derivative
$$ 12 + 13x + 14x^2 dx $$

We represent the polynomial as a list of coefficients, in increasing order of powers:
$$ [12, 13, 14] $$

We call the function with the polynomial and the increasing flag set to True (because the polynomial is in increasing order of powers):

The result is the derivative of the polynomial, also represented as a list of coefficients:

Let's calculate the derivative of the same polynomial, but this time represented in decreasing order of powers:
$$ 12x^2 + 13x + 14 dx $$

In [2]:
derive_poly(poly = [12, 13, 14])  # 12x^2 + 13x + 14 dx = 24x + 13

[24, 13]

if we wanted to have the list of polynomials in increasing order we can also do it as:

In [3]:
derive_poly(poly = [13, 13, 12], increasing=True)  # 12x^2 + 13x + 14 dx = 24x + 13

[13, 24]

Now let's se how the `increasing` parameter

$$ x + 2x^2 dx = 1 + 4x $$

We can try with different polynomials in both increasing and decreasing orders:

In [4]:
derive_poly([0, 1, 2], increasing=True)  # 0 + x + 2x^2. dx = 1 + 4x

[1, 4]

 Here we run get the  same result as above but reversed.

In [5]:
poly = [2, 1, 0]
result = derive_poly(poly, False)  # 2x^2 + 1x + 0 dx = 4x + 1
# asserts the function output is correct, fails otherwise
assert result == [4, 1], "incorrect derivation"  
result

[4, 1]

### Extensive testings

**Test 1: Derivative of a Constant**
Constants have a derivative of zero. 

In [6]:
# Derivative of 7 = 0
poly = [7]
result = derive_poly(poly)
print("Derivative of poly:", result)
assert result == [], "incorrect derivation"  # Expected output []

Derivative of poly: []


**Test 2: Linear Function**
For a linear function \( ax + b \), the derivative is \( a \).

In [7]:
# Derivative of 3x + 4 = 3
poly = [3, 4]
result = derive_poly(poly, increasing=False)
print("Derivative of poly:", result)
assert result == [3], "incorrect derivation"  # Expected output [3]

Derivative of poly: [3]


**Test 3: Quadratic Function**
A quadratic function \( ax^2 + bx + c \) will have a linear derivative.

$$
ax^2+bx+c
$$

will have a linear derivative.

$$
2ax+b
$$

In [8]:
# Derivative of 2x^2 + 4x + 1 = 4x + 4
poly = [1, 4, 2]
result = derive_poly(poly, True)
print("Derivative of poly:", result)
assert result == [4, 4], "incorrect derivation"  # Expected output [4, 4]

Derivative of poly: [4, 4]


**Test 4: Cubic Function**
For a cubic function \( ax^3 + bx^2 + cx + d \), the derivative will be quadratic.

In [9]:
# Derivative of 4 + 3x + 2x^2 + x^3 = 3 + 4x + 3x^2
poly = [4, 3, 2, 1]
result = derive_poly(poly, increasing=True)
print("Derivative of poly:", result)
assert result == [3, 4, 3], "incorrect derivation"  # Expected output [3, 4, 3]

Derivative of poly: [3, 4, 3]


**Test 5: Derivative with Zero Coefficients**
For \( x^3 + 0x^2 + 2x + 3 \), zero coefficients should be correctly handled.

In [10]:
# Derivative of x^3 + 0x^2 + 2x + 3 = 2 + 0x + 3x^2
poly = [3, 2, 0, 1]
result = derive_poly(poly, increasing=True)
print("Derivative of poly:", result)
assert result == [2, 0, 3], "incorrect derivation"  # Expected output [2, 0, 3]

Derivative of poly: [2, 0, 3]


These test cases can help us ensure that the `derive_poly` function is working correctly across different types of polynomials.

---

## evaluate_polynomial

We can create a set of test cases to validate the behavior of `evaluate_polynomial` function. Below are some tests to cover different scenarios.

In [11]:
# Example usage:
coefficients = (2, -1, 3)  # Represents the polynomial 2x^2 - x + 3
x_value = 4.0
result = evaluate_polynomial(x_value, *coefficients)
print(f"The result of evaluating the polynomial at x = {x_value} is {result}")

The result of evaluating the polynomial at x = 4.0 is 31.0


#### Example usage:
Let's evaluate the polynomial
$$ 12 + 13x + 14x^2 $$
at the value
$$ x = 4 $$

In [12]:
coefficients = (14, 13, 12)
x_value = 4.0

result = evaluate_polynomial(x_value, *coefficients)
print(f"The result of evaluating the polynomial at x = {x_value} is {result}")

The result of evaluating the polynomial at x = 4.0 is 288.0


The result is the value of the polynomial at the given value of x:

$$ 12 + 13x + 14x^2 = 12 + 13*4 + 14*4^2 = 12 + 52 + 224 = 288 $$

**Test 1: Evaluating a Constant Polynomial**

Let's run an example with a polynomial of degree 0:
$$ 12 $$

The result is the value of the polynomial at the given value of x:

$$ 12 = 12 $$

In [13]:
coefficients = (12,)
x_value = 4.0

result = evaluate_polynomial(x_value, *coefficients)
print(f"The result of evaluating the polynomial at x = {x_value} is {result}")

The result of evaluating the polynomial at x = 4.0 is 12.0


For a constant polynomial like \( f(x) = 7 \), the function should return 7 for any value of \( x \).

In [14]:
coefficients = (7,)
x = 2.0
result = evaluate_polynomial(x, *coefficients)
print(f"The result of evaluating the polynomial at x = {x} is {result}")
assert result == 7, "incorrect evaluation"  # 7

The result of evaluating the polynomial at x = 2.0 is 7.0


**Test 2: Evaluating a Linear Polynomial**
Test a linear polynomial like \( f(x=30) = 4x + 4x= 123  \).

In [15]:
coefficients = (4, 3)  # 4(x) + 3
x = 30.0
result = evaluate_polynomial(x, *coefficients)
print(f"The result of evaluating the polynomial at x = {x} is {result}")
assert result == 123, "incorrect evaluation"  # 4(30) + 3 = 123 

The result of evaluating the polynomial at x = 30.0 is 123.0


**Test 3: Evaluating a Quadratic Polynomial with Zero Coefficients**
Test a quadratic polynomial like \( f(x=2) = 4x^2 + 3x + 1 = 17 \).
Polynomials may contain terms with zero coefficients, which should be handled correctly.

In [16]:
coefficients = (4, 0, 1)  # 4x^2 + 1
x_value = 2.0
result = evaluate_polynomial(x_value, *coefficients)
print(f"The result of evaluating the polynomial at x = {x_value} is {result}")
assert result == 17, "incorrect evaluation"  # 4(2)^2 + 1 = 17

The result of evaluating the polynomial at x = 2.0 is 17.0


**Test 4: Evaluating a beyond cubic Polynomial**
For a cubic polynomial like \( f(x) = x^4 - 3x^3 - 2x^2 + 1x + 0 \).

In [17]:
coefficients = (1, -3, 2, 1, 0)
x = 3.0
result = evaluate_polynomial(x, *coefficients)
print(f"The result of evaluating the polynomial at x = {x} is {result}")
assert result == 21, "incorrect evaluation"  # 1(x)^4 - 3(x)^3 + 2(x)^2 + 1(x) + 0 = 21

The result of evaluating the polynomial at x = 3.0 is 21.0


In [18]:
coefficients = (1, 1, 1, 1, 1, 0)
x = 5.0
result = evaluate_polynomial(x, *coefficients)
print(f"The result of evaluating the polynomial at x = {x} is {result}")
assert result == 3905, "incorrect evaluation"  # x^5 + x^4 + x^3 + x^2 + x = 3905

The result of evaluating the polynomial at x = 5.0 is 3905.0


**Test 5: Evaluating mixed cases**

In [19]:
coefficients = (0, -6, 5, -4, 3, -2, 1, 0)
x = 10.0
result = evaluate_polynomial(x, *coefficients)
print(f"The result of evaluating the polynomial at x = {x} is {result}")
assert result == -5537190, "incorrect evaluation"  # 0x^7 - 6x^6 + 5x^5 - 4x^4 + 3x^3 - 2x^2 + x + 0 = -5537190

The result of evaluating the polynomial at x = 10.0 is -5537190.0


These test cases should help validate that the `evaluate_polynomial` function works correctly across different types of polynomials and input values.

---

# 2: Integral

## Symbolical integration with substitution
Below are some test cases that cover a variety of scenarios for the numerical integration function `integral`.

**Test 1: Constant Polynomial**
This is to test the simplest case, where the polynomial is a constant term like $$ f(x) = 7 $$

In [20]:
Poly = {0: 7}  # 7
A = 0
B = 2
result = integral(Poly, A, B)  # Expected output: 14
print("The integral from {} to {} is: {}".format(A, B, result))
assert abs(result - 14) < 1e-9, "incorrect integration"

The integral from 0 to 2 is: 14.0


**Test 2: Linear Polynomial**
A test case for a linear polynomial $$ f(x) = 4x - 1 $$

In [21]:
Poly = {1: 4, 0: -1}  # 4x - 1
A = 1
B = 2
result = integral(Poly, A, B)  # Expected output: 5
print("The integral from {} to {} is: {}".format(A, B, result))
assert abs(result - 5) < 1e-9, "incorrect integration"

The integral from 1 to 2 is: 5.0


**Test 3: Non-zero Coefficients for Even and Odd Powers**
This test is for a polynomial like $$f(x) = 2*x^{3} - 3*x^{2} + 4x + 6$$

In [22]:
Poly = {3: 2, 2: -3, 1: 4, 0: 6}  # 2x^3 - 3x^2 + 4x + 6
A = 0
B = 1
result = integral(Poly, A, B)  # Expected output: 7.5
print("The integral from {} to {} is: {}".format(A, B, result))
assert abs(result - 7.5) < 1e-9, "incorrect integration"

The integral from 0 to 1 is: 7.5


**Test 4: Polynomial with Zero Coefficients**
A polynomial like $$f(x) = x^4 + 0x^2 + 2$$

In [23]:
Poly = {4: 1, 2: 0, 0: 2}  # x^4 + 2
A = 0
B = 1
result = integral(Poly, A, B)  # Expected output: 2.2
print("The integral from {} to {} is: {}".format(A, B, result))
assert abs(result - 2.2) < 1e-9, "incorrect integration"

The integral from 0 to 1 is: 2.2


**Test 5: Polynomials with Fractional Coefficients**

To test the precision of your integration function, you could use fractional coefficients. For instance, consider
$$
f(x) = (3/4)x^2 + (1/2)x + (1/4)
$$

In [24]:
Poly = {2: 0.75, 1: 0.5, 0: 0.25}  # 0.75x^2 + 0.5x + 0.25
A = 0
B = 1
result = integral(Poly, A, B)  # Expected output: 0.75
print("The integral from {} to {} is: {}".format(A, B, result))
assert abs(result - 0.75) < 1e-9, "incorrect integration"

The integral from 0 to 1 is: 0.75


These tests should provide comprehensive coverage for our numerical integration function, helping ensure its accuracy and robustness.

## Riemann Integral Approximation

#### Edge Case 1: Zero Polynomial

$$ ∫ 0 = 0 dx $$
x in the interval \([0, 5]\):

$$ = 0 $$

In [25]:
poly = [(0, 0)]
step = 0.01
start = 0
stop = 5
result = riemann_integral_approximation(poly, step, start, stop)
assert result == 0, "Result is not 0."
print(f'Numerical Result: {result}')

Numerical Result: 0.0


#### Edge Case 2: Negative Coefficients:
Testing a polynomial with negative coefficients.
> ∫-2x^2 - 3x - 4, x = [1,3]

$$ ∫ -2x^2 - 3x - 4 $$
$$ = -\frac{2}{3}x^3  -  \frac{3}{2}x^2 - 4x  dx $$

x in the interval [1, 3]\:

$$
= (-\frac{2}{3}x^3 - \frac{3}{2}(3)^2 - 4(3)) - (-\frac{2}{3}(1)^3 - \frac{3}{2}(1)^2 - 4(1))
$$

$$ = -37.33 $$


In [26]:
poly = [(-2, 2), (-3, 1), (-4, 0)]
step = 0.01
start = 1
stop = 3
result = riemann_integral_approximation(poly, step, start, stop)
print(f'Numerical Result: {result}')
assert (result + 37.33) < 0.5, "Result is not within 0.5 of the expected value."

Numerical Result: -37.22340000000003


#### Edge Case 3: Non-integer Coefficients and Powers:
Testing a polynomial with non-integer coefficients and powers.

$$
∫ 1.5x^{1.5} + 2.5x^{1.5} = x^{2.5} + \frac{5}{3}x^{1.5} dx
$$

In [27]:
poly = [(1.5, 1.5), (-2.5, 1.5)]
step = 0.01
start = 1
stop = 4
result = riemann_integral_approximation(poly, step, start, stop)
print(f'Numerical Result: {result}')
assert abs(12.4 + result) < 0.5, "Result is not within 0.5 of the expected value."

Numerical Result: -12.365012499995458


#### Edge Case 4: Non-integer Interval:

$$ x^{-1} + x^{-2} = ln|x| - x^{-1}  dx $$
x in the interval [1, 2]:

$$ = ln|2| - 2^{-1} - ln|1| - 1^{-1}  $$

$$ = ln(2) + 1 1.693 $$


In [28]:
poly = [(1, -1), (1, -2)]
step = 0.0001
start = 1
stop = 2
result = riemann_integral_approximation(poly, step, start, stop)
print(f'Numerical Result: {result}')
assert abs(result - 1.19315) < 0.05, "Result is not within 0.05 of the expected value."

Numerical Result: 1.1932096826433303


## Reflection

I'm pleased with how my code turned out overall. I tried to structure it in a modular way, with each function focused on a specific task. The functions drop_leftmost_zeros and derive_poly work together nicely to handle removing leading zeros from polynomial derivatives. I think my use of helper functions like drop_leftmost_zeros made the main derive_poly function easier to read and understand.

When writing evaluate_polynomial, I was careful to use tuple unpacking and a generator expression rather than temporary variables, which kept it clean and concise. I also tried to use appropriate data structures like lists and dicts to represent mathematical concepts in a natural way. For example, representing polynomials as lists or dicts maps well to how we think about polynomials.

There are some areas I could improve on though. For integral, I probably could have structured it more similarly to evaluate_polynomial using a generator expression instead of a simpler sum. That would make it more consistent with the other functions. I also should add some type hints and input validation to make the functions easier to use correctly and safely. And I need to write more tests - that will help me catch bugs and edge cases.

Overall, writing this code taught me the importance of modularity, picking the right data structures, and writing readable code. I'm glad I asked others for feedback on my code - getting suggestions for improvement from peers has really helped strengthen my coding skills. This experience has shown me how openness to critiques allows me to keep improving as a programmer. I'm excited to continue honing my skills as I take on more coding challenges!

#### Usage of peer code
I haven't seen any code from my peers that I could use in my code at the time of this submission.